File: yeet-synth/synth/scenarios.py
"""
User behaviour scenarios.

Each scenario models a realistic sequence of API calls made by
a particular type of user. Scenarios are the primary unit of
traffic composition — the runner picks scenarios by weight.

Archetypes:
  anonymous        — unauthenticated probing (health, warmup)
  authenticated    — login → profile → config → logout
  active_bettor    — full game flow: session → bets → close
  wallet_heavy     — deposit → check balance → withdraw → history
  admin            — internal diagnostic endpoints
"""

from __future__ import annotations

import asyncio
import logging
import random
from dataclasses import dataclass
from typing import Optional

import synth.endpoints as ep

from .client import SynthClient
from .payloads import FLAG_KEYS
from .token_manager import TokenPool

logger = logging.getLogger(__name__)


@dataclass
class ScenarioResult:
    scenario: str
    requests: int = 0
    successes: int = 0
    errors: int = 0
    skipped: bool = False
    skip_reason: str = ""

    @property
    def success_rate(self) -> float:
        return self.successes / self.requests if self.requests > 0 else 0.0


async def _think(min_ms: float = 50, max_ms: float = 300) -> None:
    """Simulate inter-request think time."""
    await asyncio.sleep(random.uniform(min_ms, max_ms) / 1000.0)


class AnonymousScenario:
    """
    Unauthenticated traffic: health probes and warm-path checks.
    Models load balancers, monitoring agents, and pre-auth browsers.
    """

    name = "anonymous"

    async def run(self, client: SynthClient, pool: TokenPool) -> ScenarioResult:
        result = ScenarioResult(scenario=self.name)

        checks = [ep.health_live, ep.health_ready, ep.health_dependencies]
        for fn in checks:
            r = await fn(client)
            result.requests += 1
            if 200 <= r.status_code < 300:
                result.successes += 1
            else:
                result.errors += 1
            await _think(10, 80)

        return result


class AuthenticatedUserScenario:
    """
    Standard authenticated user flow:
    login → validate session → fetch profile → check limits → fetch flags → logout

    Represents casual users checking account status or browsing.
    """

    name = "authenticated"

    async def run(self, client: SynthClient, pool: TokenPool) -> ScenarioResult:
        result = ScenarioResult(scenario=self.name)
        creds = pool.get_random()
        if not creds:
            result.skipped = True
            result.skip_reason = "empty token pool"
            return result

        await pool.maybe_refresh(creds)
        token = creds.access_token
        user_id = creds.user_id

        steps = [
            lambda: ep.auth_session_validate(client, token),
            lambda: ep.users_profile(client, token, user_id),
            lambda: ep.users_limits(client, token, user_id),
            lambda: ep.config_flags(client, token),
            lambda: ep.config_flag(client, token, random.choice(FLAG_KEYS)),
        ]
        for step in steps:
            r = await step()
            result.requests += 1
            if 200 <= r.status_code < 300:
                result.successes += 1
            elif r.status_code == 404:
                result.successes += 1  # 404 on a flag key is expected
            else:
                result.errors += 1
            await _think(100, 500)

        return result


class ActiveBettorScenario:
    """
    Full gaming flow — the highest-traffic, most critical path:
    balance → create session → [N bets with heartbeats] → close session → history

    This scenario exercises the entire transactional stack:
    wallet reservation, risk evaluation, bet settlement.
    """

    name = "active_bettor"

    def __init__(self, bets_per_session: int = 3) -> None:
        self.bets_per_session = bets_per_session

    async def run(self, client: SynthClient, pool: TokenPool) -> ScenarioResult:
        result = ScenarioResult(scenario=self.name)
        creds = pool.get_random()
        if not creds:
            result.skipped = True
            result.skip_reason = "empty token pool"
            return result

        await pool.maybe_refresh(creds)
        token = creds.access_token
        user_id = creds.user_id

        # Pre-flight: check balance
        r = await ep.wallet_balance(client, token, user_id)
        result.requests += 1
        result.successes += 1 if r.status_code == 200 else 0
        result.errors += 0 if r.status_code == 200 else 1
        await _think(50, 150)

        # Create game session
        r = await ep.game_session_create(client, token)
        result.requests += 1
        session_id: Optional[str] = None
        if r.status_code == 201:
            result.successes += 1
            # We can't easily get body from RequestRecord in the current design;
            # generate a plausible session_id for downstream calls and handle
            # 404s gracefully. In a full impl, client would return body too.
            session_id = None  # downstream bets will omit session_id
        else:
            result.errors += 1
        await _think(100, 300)

        # Place bets
        for i in range(self.bets_per_session):
            r = await ep.bet_place(client, token, session_id=session_id)
            result.requests += 1
            if r.status_code in (200, 201, 202):
                result.successes += 1
            elif r.status_code == 402:
                # Insufficient funds — not an error in synthetic context
                result.successes += 1
            else:
                result.errors += 1

            # Heartbeat every other bet
            if session_id and i % 2 == 1:
                await ep.game_session_heartbeat(client, token, session_id)
                result.requests += 1
                result.successes += 1
            await _think(200, 800)

        # Fetch bet history
        r = await ep.bet_history(client, token, limit=10)
        result.requests += 1
        result.successes += 1 if r.status_code == 200 else 0
        result.errors += 0 if r.status_code == 200 else 1
        await _think(100, 300)

        # Close session if we have one
        if session_id:
            r = await ep.game_session_close(client, token, session_id)
            result.requests += 1
            result.successes += 1 if r.status_code == 200 else 0
            result.errors += 0 if r.status_code == 200 else 1

        return result


class WalletHeavyScenario:
    """
    Wallet-focused flow: deposit → check balance → transaction history
    → optional withdraw. Models payment-heavy users (cashiers, finance
    validation).
    """

    name = "wallet_heavy"

    async def run(self, client: SynthClient, pool: TokenPool) -> ScenarioResult:
        result = ScenarioResult(scenario=self.name)
        creds = pool.get_random()
        if not creds:
            result.skipped = True
            result.skip_reason = "empty token pool"
            return result

        await pool.maybe_refresh(creds)
        token = creds.access_token
        user_id = creds.user_id

        # Deposit
        r = await ep.wallet_deposit(client, token, user_id, amount="100.00")
        result.requests += 1
        result.successes += 1 if r.status_code in (200, 201) else 0
        result.errors += 0 if r.status_code in (200, 201) else 1

        # Idempotency replay test — same key, should return same result
        if r.status_code in (200, 201):
            # Re-record last idempotency key; note: in our client the
            # idem key is generated per request. For a true replay test,
            # pass the same key explicitly.
            pass
        await _think(100, 200)

        # Balance check
        r = await ep.wallet_balance(client, token, user_id)
        result.requests += 1
        result.successes += 1 if r.status_code == 200 else 0
        result.errors += 0 if r.status_code == 200 else 1
        await _think(50, 100)

        # Transaction history (multiple pages)
        for _ in range(random.randint(1, 3)):
            r = await ep.wallet_transactions(client, token, user_id, limit=10)
            result.requests += 1
            result.successes += 1 if r.status_code == 200 else 0
            result.errors += 0 if r.status_code == 200 else 1
            await _think(80, 200)

        # Occasional withdraw
        if random.random() < 0.3:
            r = await ep.wallet_withdraw(client, token, user_id, amount="20.00")
            result.requests += 1
            result.successes += 1 if r.status_code in (200, 201, 202) else 0
            result.errors += 0 if r.status_code in (200, 201, 202) else 1

        # Risk signal after large wallet activity
        if random.random() < 0.5:
            r = await ep.risk_signal(client, token, user_id)
            result.requests += 1
            result.successes += 1 if r.status_code == 202 else 0
            result.errors += 0 if r.status_code == 202 else 1

        return result


class AdminScenario:
    """
    Internal diagnostic flow: status, config, db stats.
    Low frequency. Requires admin token — skips gracefully if unavailable.
    """

    name = "admin"

    async def run(self, client: SynthClient, pool: TokenPool) -> ScenarioResult:
        result = ScenarioResult(scenario=self.name)
        admin = pool.get_admin()
        if not admin:
            result.skipped = True
            result.skip_reason = "no admin token"
            return result

        token = admin.access_token
        for fn in [ep.admin_status, ep.admin_config, ep.admin_db_stats]:
            r = await fn(client, token)
            result.requests += 1
            if 200 <= r.status_code < 300:
                result.successes += 1
            else:
                result.errors += 1
            await _think(200, 500)

        return result


# ── Registry ──────────────────────────────────────────────────────────────────

SCENARIO_REGISTRY: dict[str, type] = {
    "anonymous": AnonymousScenario,
    "authenticated": AuthenticatedUserScenario,
    "active_bettor": ActiveBettorScenario,
    "wallet_heavy": WalletHeavyScenario,
    "admin": AdminScenario,
}


def pick_scenario(weights: dict[str, float]) -> type:
    """Weighted random scenario selection."""
    names = list(weights.keys())
    wts = [weights[n] for n in names]
    chosen = random.choices(names, weights=wts, k=1)[0]
    cls = SCENARIO_REGISTRY.get(chosen, AuthenticatedUserScenario)
    return cls

File: y_eet-synth/internal/scenarios/scenarios.go
// Package scenarios defines user behaviour archetypes.
// Each scenario models a realistic sequence of API calls for a particular
// type of user. The runner picks scenarios by weight and runs them concurrently.
package scenarios

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"y_eet-synth/internal/client"
	"y_eet-synth/internal/token"
)

// Result summarises the outcome of one scenario execution.
type Result struct {
	Scenario  string
	Requests  int
	Successes int
	Errors    int
	Skipped   bool
	SkipReason string
}

func (r *Result) SuccessRate() float64 {
	if r.Requests == 0 {
		return 0
	}
	return float64(r.Successes) / float64(r.Requests)
}

// ScenarioFunc is the signature for all scenario implementations.
type ScenarioFunc func(ctx context.Context, cl *client.Client, pool *token.Pool) Result

// Registry maps scenario name to its function.
var Registry = map[string]ScenarioFunc{
	"anonymous":           Anonymous,
	"authenticated":       Authenticated,
	"active_bettor":       ActiveBettor,
	"wallet_heavy":        WalletHeavy,
	"admin":               Admin,
	"registration_funnel": RegistrationFunnel,
	"high_roller":         HighRoller,
	"live_event_bettor":   LiveEventBettor,
}

// Pick selects a scenario by weighted random choice.
func Pick(weights map[string]float64) ScenarioFunc {
	total := 0.0
	for _, w := range weights {
		total += w
	}
	r := rand.Float64() * total //nolint:gosec
	acc := 0.0
	for name, w := range weights {
		acc += w
		if r < acc {
			if fn, ok := Registry[name]; ok {
				return fn
			}
		}
	}
	// fallback
	return Authenticated
}

// ── Helpers ───────────────────────────────────────────────────────────────────

func think(minMs, maxMs int) {
	ms := minMs + rand.Intn(maxMs-minMs+1) //nolint:gosec
	time.Sleep(time.Duration(ms) * time.Millisecond)
}

func ok2xx(status int) bool { return status >= 200 && status < 300 }

func amount(min, max float64) string {
	return fmt.Sprintf("%.2f", min+rand.Float64()*(max-min)) //nolint:gosec
}

var gameIDs = []string{
	"game_crash_v1", "game_crash_v2", "game_slots_classic", "game_slots_mega",
	"game_slots_turbo", "game_roulette_eu", "game_blackjack_std", "game_poker_texas",
	"game_dice_classic", "game_plinko_v1", "game_mines_v1",
}

var liveGameIDs = []string{
	"game_crash_v1", "game_crash_v2", "game_sports_live",
	"game_esports_cs", "game_esports_lol", "game_plinko_v1",
}

var flagKeys = []string{
	"risk_eval_enabled", "instant_settlement_enabled",
	"withdrawal_kyc_gate", "new_game_engine_pct", "bonus_engine_v2",
}

// ── Scenarios ─────────────────────────────────────────────────────────────────

// Anonymous models unauthenticated health probes and warmup checks.
func Anonymous(ctx context.Context, cl *client.Client, pool *token.Pool) Result {
	res := Result{Scenario: "anonymous"}
	paths := []string{"/health/live", "/health/ready", "/health/dependencies"}
	names := []string{
		"GET /health/live", "GET /health/ready", "GET /health/dependencies",
	}
	for i, path := range paths {
		r := cl.Do(ctx, client.RequestOptions{
			Method: "GET", Path: path, EndpointName: names[i],
		})
		res.Requests++
		if ok2xx(r.StatusCode) {
			res.Successes++
		} else {
			res.Errors++
		}
		think(10, 80)
	}
	return res
}

// Authenticated models a standard logged-in user browsing their account.
func Authenticated(ctx context.Context, cl *client.Client, pool *token.Pool) Result {
	res := Result{Scenario: "authenticated"}
	creds := pool.GetRandom()
	if creds == nil {
		res.Skipped = true
		res.SkipReason = "empty token pool"
		return res
	}
	pool.MaybeRefresh(ctx, creds)
	tok := creds.AccessToken
	uid := creds.UserID

	type step struct {
		method string
		path   string
		name   string
	}
	steps := []step{
		{"GET", "/api/v1/auth/session/validate", "GET /api/v1/auth/session/validate"},
		{"GET", "/api/v1/users/" + uid + "/profile", "GET /api/v1/users/:id/profile"},
		{"GET", "/api/v1/users/" + uid + "/limits", "GET /api/v1/users/:id/limits"},
		{"GET", "/api/v1/config/flags", "GET /api/v1/config/flags"},
		{"GET", "/api/v1/config/flags/" + flagKeys[rand.Intn(len(flagKeys))], "GET /api/v1/config/flags/:key"}, //nolint:gosec
	}
	for _, s := range steps {
		r := cl.Do(ctx, client.RequestOptions{
			Method: s.method, Path: s.path, Token: tok, EndpointName: s.name,
		})
		res.Requests++
		if ok2xx(r.StatusCode) || r.StatusCode == 404 {
			res.Successes++
		} else {
			res.Errors++
		}
		think(100, 500)
	}
	return res
}

// ActiveBettor models the full betting flow: balance → session → bets → close.
func ActiveBettor(ctx context.Context, cl *client.Client, pool *token.Pool) Result {
	res := Result{Scenario: "active_bettor"}
	creds := pool.GetRandom()
	if creds == nil {
		res.Skipped = true
		res.SkipReason = "empty token pool"
		return res
	}
	pool.MaybeRefresh(ctx, creds)
	tok := creds.AccessToken
	uid := creds.UserID

	// balance check
	r := cl.Do(ctx, client.RequestOptions{
		Method: "GET", Path: "/api/v1/wallet/" + uid + "/balance",
		Token: tok, EndpointName: "GET /api/v1/wallet/:id/balance",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	think(50, 150)

	// create session
	r = cl.Do(ctx, client.RequestOptions{
		Method: "POST", Path: "/api/v1/games/sessions",
		Token: tok, IdempotencyKey: newUUID(),
		Body:         createSessionBody(gameIDs[rand.Intn(len(gameIDs))]), //nolint:gosec
		EndpointName: "POST /api/v1/games/sessions",
	})
	res.Requests++
	if r.StatusCode == 201 {
		res.Successes++
	} else {
		res.Errors++
	}
	think(100, 300)

	// place bets
	bets := 5 + rand.Intn(5) //nolint:gosec
	for i := 0; i < bets; i++ {
		r = cl.Do(ctx, client.RequestOptions{
			Method: "POST", Path: "/api/v1/bets/place",
			Token: tok, IdempotencyKey: newUUID(),
			Body:         placeBetBody("", ""),
			EndpointName: "POST /api/v1/bets/place",
		})
		res.Requests++
		if ok2xx(r.StatusCode) || r.StatusCode == 402 {
			res.Successes++
		} else {
			res.Errors++
		}
		think(200, 800)
	}

	// history
	r = cl.Do(ctx, client.RequestOptions{
		Method: "GET", Path: "/api/v1/bets/history",
		Token: tok, Params: map[string]string{"limit": "10"},
		EndpointName: "GET /api/v1/bets/history",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	return res
}

// WalletHeavy models a payment-focused user: deposit → balance → history → optional withdraw.
func WalletHeavy(ctx context.Context, cl *client.Client, pool *token.Pool) Result {
	res := Result{Scenario: "wallet_heavy"}
	creds := pool.GetRandom()
	if creds == nil {
		res.Skipped = true
		res.SkipReason = "empty token pool"
		return res
	}
	pool.MaybeRefresh(ctx, creds)
	tok := creds.AccessToken
	uid := creds.UserID

	r := cl.Do(ctx, client.RequestOptions{
		Method: "POST", Path: "/api/v1/wallet/" + uid + "/deposit",
		Token: tok, IdempotencyKey: newUUID(),
		Body:         depositBody("100.00"),
		EndpointName: "POST /api/v1/wallet/:id/deposit",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	think(100, 200)

	r = cl.Do(ctx, client.RequestOptions{
		Method: "GET", Path: "/api/v1/wallet/" + uid + "/balance",
		Token: tok, EndpointName: "GET /api/v1/wallet/:id/balance",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	think(50, 100)

	pages := 1 + rand.Intn(3) //nolint:gosec
	for i := 0; i < pages; i++ {
		r = cl.Do(ctx, client.RequestOptions{
			Method: "GET", Path: "/api/v1/wallet/" + uid + "/transactions",
			Token: tok, Params: map[string]string{"limit": "10"},
			EndpointName: "GET /api/v1/wallet/:id/transactions",
		})
		res.Requests++
		if ok2xx(r.StatusCode) {
			res.Successes++
		} else {
			res.Errors++
		}
		think(80, 200)
	}

	if rand.Float64() < 0.3 { //nolint:gosec
		r = cl.Do(ctx, client.RequestOptions{
			Method: "POST", Path: "/api/v1/wallet/" + uid + "/withdraw",
			Token: tok, IdempotencyKey: newUUID(),
			Body:         withdrawBody("20.00"),
			EndpointName: "POST /api/v1/wallet/:id/withdraw",
		})
		res.Requests++
		if ok2xx(r.StatusCode) {
			res.Successes++
		} else {
			res.Errors++
		}
	}

	if rand.Float64() < 0.5 { //nolint:gosec
		r = cl.Do(ctx, client.RequestOptions{
			Method: "POST", Path: "/api/v1/risk/signals",
			Token: tok, Body: riskSignalBody(uid),
			EndpointName: "POST /api/v1/risk/signals",
		})
		res.Requests++
		if r.StatusCode == 202 {
			res.Successes++
		} else {
			res.Errors++
		}
	}

	return res
}

// Admin models internal diagnostic calls.
func Admin(ctx context.Context, cl *client.Client, pool *token.Pool) Result {
	res := Result{Scenario: "admin"}
	admin := pool.GetAdmin()
	if admin == nil {
		res.Skipped = true
		res.SkipReason = "no admin token"
		return res
	}
	tok := admin.AccessToken

	for _, ep := range []struct{ path, name string }{
		{"/_internal/status", "GET /_internal/status"},
		{"/_internal/config", "GET /_internal/config"},
		{"/_internal/db/stats", "GET /_internal/db/stats"},
	} {
		r := cl.Do(ctx, client.RequestOptions{
			Method: "GET", Path: ep.path, Token: tok,
			Internal: true, EndpointName: ep.name,
		})
		res.Requests++
		if ok2xx(r.StatusCode) {
			res.Successes++
		} else {
			res.Errors++
		}
		think(200, 500)
	}
	return res
}

// RegistrationFunnel models new-user onboarding: register → profile → deposit → first bet.
func RegistrationFunnel(ctx context.Context, cl *client.Client, pool *token.Pool) Result {
	res := Result{Scenario: "registration_funnel"}
	// Use the pool's http client by making a direct request; we need the response body.
	// For simplicity, skip to the first available pool user and simulate the flow.
	creds := pool.GetRandom()
	if creds == nil {
		res.Skipped = true
		res.SkipReason = "empty token pool"
		return res
	}
	pool.MaybeRefresh(ctx, creds)
	tok := creds.AccessToken
	uid := creds.UserID

	for _, s := range []struct{ method, path, name string }{
		{"GET", "/api/v1/users/" + uid + "/profile", "GET /api/v1/users/:id/profile"},
		{"GET", "/api/v1/users/" + uid + "/limits", "GET /api/v1/users/:id/limits"},
		{"GET", "/api/v1/config/flags", "GET /api/v1/config/flags"},
	} {
		r := cl.Do(ctx, client.RequestOptions{
			Method: s.method, Path: s.path, Token: tok, EndpointName: s.name,
		})
		res.Requests++
		if ok2xx(r.StatusCode) {
			res.Successes++
		} else {
			res.Errors++
		}
		think(300, 800)
	}

	r := cl.Do(ctx, client.RequestOptions{
		Method: "POST", Path: "/api/v1/wallet/" + uid + "/deposit",
		Token: tok, IdempotencyKey: newUUID(),
		Body: depositBody(amount(20, 150)), EndpointName: "POST /api/v1/wallet/:id/deposit",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	think(600, 2000)

	r = cl.Do(ctx, client.RequestOptions{
		Method: "POST", Path: "/api/v1/games/sessions",
		Token: tok, IdempotencyKey: newUUID(),
		Body: createSessionBody(""), EndpointName: "POST /api/v1/games/sessions",
	})
	res.Requests++
	if r.StatusCode == 201 {
		res.Successes++
	} else {
		res.Errors++
	}
	think(800, 2000)

	bets := 1 + rand.Intn(3) //nolint:gosec
	for i := 0; i < bets; i++ {
		r = cl.Do(ctx, client.RequestOptions{
			Method: "POST", Path: "/api/v1/bets/place",
			Token: tok, IdempotencyKey: newUUID(),
			Body: placeBetBody("", amount(1, 15)), EndpointName: "POST /api/v1/bets/place",
		})
		res.Requests++
		if ok2xx(r.StatusCode) || r.StatusCode == 402 {
			res.Successes++
		} else {
			res.Errors++
		}
		think(500, 1500)
	}
	return res
}

// HighRoller models a high-value bettor with large deposits and rapid bets.
func HighRoller(ctx context.Context, cl *client.Client, pool *token.Pool) Result {
	res := Result{Scenario: "high_roller"}
	creds := pool.GetRandom()
	if creds == nil {
		res.Skipped = true
		res.SkipReason = "empty token pool"
		return res
	}
	pool.MaybeRefresh(ctx, creds)
	tok := creds.AccessToken
	uid := creds.UserID

	r := cl.Do(ctx, client.RequestOptions{
		Method: "POST", Path: "/api/v1/wallet/" + uid + "/deposit",
		Token: tok, IdempotencyKey: newUUID(),
		Body: depositBody(amount(500, 5000)), EndpointName: "POST /api/v1/wallet/:id/deposit",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	think(80, 200)

	r = cl.Do(ctx, client.RequestOptions{
		Method: "GET", Path: "/api/v1/wallet/" + uid + "/balance",
		Token: tok, EndpointName: "GET /api/v1/wallet/:id/balance",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	think(50, 100)

	bets := 10 + rand.Intn(16) //nolint:gosec
	for i := 0; i < bets; i++ {
		r = cl.Do(ctx, client.RequestOptions{
			Method: "POST", Path: "/api/v1/bets/place",
			Token: tok, IdempotencyKey: newUUID(),
			Body: placeBetBody("", amount(50, 1000)), EndpointName: "POST /api/v1/bets/place",
		})
		res.Requests++
		if ok2xx(r.StatusCode) || r.StatusCode == 402 {
			res.Successes++
		} else {
			res.Errors++
		}
		if i > 0 && i%5 == 0 {
			r2 := cl.Do(ctx, client.RequestOptions{
				Method: "POST", Path: "/api/v1/risk/signals",
				Token: tok, Body: riskSignalBody(uid),
				EndpointName: "POST /api/v1/risk/signals",
			})
			res.Requests++
			if r2.StatusCode == 202 {
				res.Successes++
			} else {
				res.Errors++
			}
		}
		think(80, 350)
	}

	if rand.Float64() < 0.65 { //nolint:gosec
		r = cl.Do(ctx, client.RequestOptions{
			Method: "POST", Path: "/api/v1/wallet/" + uid + "/withdraw",
			Token: tok, IdempotencyKey: newUUID(),
			Body: withdrawBody(amount(200, 2000)), EndpointName: "POST /api/v1/wallet/:id/withdraw",
		})
		res.Requests++
		if ok2xx(r.StatusCode) {
			res.Successes++
		} else {
			res.Errors++
		}
	}
	return res
}

// LiveEventBettor models a rapid-fire in-play bettor.
func LiveEventBettor(ctx context.Context, cl *client.Client, pool *token.Pool) Result {
	res := Result{Scenario: "live_event_bettor"}
	creds := pool.GetRandom()
	if creds == nil {
		res.Skipped = true
		res.SkipReason = "empty token pool"
		return res
	}
	pool.MaybeRefresh(ctx, creds)
	tok := creds.AccessToken
	uid := creds.UserID
	game := liveGameIDs[rand.Intn(len(liveGameIDs))] //nolint:gosec

	r := cl.Do(ctx, client.RequestOptions{
		Method: "GET", Path: "/api/v1/wallet/" + uid + "/balance",
		Token: tok, EndpointName: "GET /api/v1/wallet/:id/balance",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	think(20, 80)

	cl.Do(ctx, client.RequestOptions{
		Method: "POST", Path: "/api/v1/games/sessions",
		Token: tok, IdempotencyKey: newUUID(),
		Body: createSessionBody(game), EndpointName: "POST /api/v1/games/sessions",
	})
	res.Requests++
	res.Successes++

	bets := 5 + rand.Intn(16) //nolint:gosec
	for i := 0; i < bets; i++ {
		r = cl.Do(ctx, client.RequestOptions{
			Method: "POST", Path: "/api/v1/bets/place",
			Token: tok, IdempotencyKey: newUUID(),
			Body: placeBetBody("", amount(5, 150)), EndpointName: "POST /api/v1/bets/place",
		})
		res.Requests++
		if ok2xx(r.StatusCode) || r.StatusCode == 402 {
			res.Successes++
		} else {
			res.Errors++
		}
		think(30, 180)
	}

	r = cl.Do(ctx, client.RequestOptions{
		Method: "GET", Path: "/api/v1/bets/history",
		Token: tok, Params: map[string]string{"limit": "20"},
		EndpointName: "GET /api/v1/bets/history",
	})
	res.Requests++
	if ok2xx(r.StatusCode) {
		res.Successes++
	} else {
		res.Errors++
	}
	return res
}

// ── Payload helpers ───────────────────────────────────────────────────────────

func createSessionBody(gameID string) map[string]interface{} {
	if gameID == "" {
		gameID = gameIDs[rand.Intn(len(gameIDs))] //nolint:gosec
	}
	return map[string]interface{}{
		"game_id":    gameID,
		"client_ref": newUUID(),
	}
}

func placeBetBody(sessionID, amt string) map[string]interface{} {
	if amt == "" {
		amt = amount(1, 200)
	}
	betTypes := []string{"spin", "straight", "auto_cashout", "split"}
	body := map[string]interface{}{
		"game_id":          gameIDs[rand.Intn(len(gameIDs))], //nolint:gosec
		"bet_type":         betTypes[rand.Intn(len(betTypes))],  //nolint:gosec
		"amount":           amt,
		"currency":         "USD",
		"client_reference": newUUID(),
	}
	if sessionID != "" {
		body["session_id"] = sessionID
	}
	return body
}

func depositBody(amt string) map[string]interface{} {
	methods := []string{"card", "crypto", "bank_transfer"}
	return map[string]interface{}{
		"amount":         amt,
		"currency":       "USD",
		"payment_method": methods[rand.Intn(len(methods))], //nolint:gosec
	}
}

func withdrawBody(amt string) map[string]interface{} {
	return map[string]interface{}{
		"amount":   amt,
		"currency": "USD",
		"method":   "bank_transfer",
	}
}

func riskSignalBody(userID string) map[string]interface{} {
	signals := []string{
		"rapid_bet_sequence", "velocity_breach", "unusual_withdrawal",
	}
	return map[string]interface{}{
		"user_id":     userID,
		"signal_type": signals[rand.Intn(len(signals))], //nolint:gosec
		"metadata":    map[string]interface{}{},
	}
}

func newUUID() string {
	b := make([]byte, 16)
	rand.Read(b) //nolint:gosec,errcheck
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:])
}
